aptos-move/aptos-gas/src/algebra.rs

```diff
@@ -0,0 +1,226 @@
+///
+/// A real life example: 3 m/s * 5 s = 15 m
+pub struct UnitDiv<U1, U2> {
+    phantom: PhantomData<(U1, U2)>,
+}
+
+/***************************************************************************************************
+ * Gas Quantities
+ *
+ **************************************************************************************************/
+/// An opqaue representation of a certain quantity, with the unit being encoded in the type.
+/// This type implements checked addition and subtraction, and only permits type-safe multiplication.
+#[derive(Serialize, Deserialize)]
+pub struct GasQuantity<U> {
+    val: u64,
+    phantom: PhantomData<U>,
+}
+
+pub type Gas = GasQuantity<GasUnit>;
+
+pub type InternalGas = GasQuantity<InternalGasUnit>;
+
+pub type NumBytes = GasQuantity<Byte>;
+
+pub type InternalGasPerByte = GasQuantity<UnitDiv<InternalGasUnit, Byte>>;
+
+pub type InternalGasPerAbstractMemoryUnit =
+    GasQuantity<UnitDiv<InternalGasUnit, AbstractMemoryUnit>>;
+
+pub type AbstractMemorySize = GasQuantity<AbstractMemoryUnit>;
+
+pub type GasScalingFactor = GasQuantity<UnitDiv<InternalGasUnit, GasUnit>>;
+
+pub type Fee = GasQuantity<Octa>;
+
+pub type FeePerGasUnit = GasQuantity<UnitDiv<Octa, GasUnit>>;
+
+/***************************************************************************************************
+ * Constructors
+ *
+ **************************************************************************************************/
+impl<U> GasQuantity<U> {
+    pub fn new(val: u64) -> Self {
+        Self {
+            val,
+            phantom: PhantomData,
+        }
+    }
+
+    pub fn zero() -> Self {
+        Self::new(0)
+    }
+}
+
+/***************************************************************************************************
+ * Conversion
+ *
+ **************************************************************************************************/
+impl<U> From<u64> for GasQuantity<U> {
+    fn from(val: u64) -> Self {
+        Self::new(val)
+    }
+}
+
+// TODO(Gas): This allows the gas value to escape the monad, which weakens the type-level
+//            protection it provides. It's currently needed for practical reasons but
+//            we should look for ways to get rid of it.
+impl<U> From<GasQuantity<U>> for u64 {
+    fn from(gas: GasQuantity<U>) -> Self {
+        gas.val
+    }
+}
+
+/***************************************************************************************************
+ * Clone & Copy
+ *
+ **************************************************************************************************/
+impl<U> Clone for GasQuantity<U> {
+    fn clone(&self) -> Self {
+        Self::new(self.val)
+    }
+}
+
+impl<U> Copy for GasQuantity<U> {}
+
+/***************************************************************************************************
+ * Display & Debug
+ *
+ **************************************************************************************************/
+impl<U> Display for GasQuantity<U> {
```

Comment:
This could be specified to give output like `5o`. Also don't you need it for `DivUnit` as well?

Reply:
> This could be specified to give output like 5o.

Sorry I'm not sure what you mean by this. Could you elaborate?

> Also don't you need it for DivUnit as well?

You mean having a custom Display impl for DivUnit that looks nice? That may be a good diea